src/main/kotlin/org/rust/lang/core/completion/RsCommonCompletionProvider.kt

```diff
@@ -228,29 +233,71 @@ object RsCommonCompletionProvider : RsCompletionProvider() {
                 result.addAllElements(variants)
             }
 
-            val lookupElement = createLookupElement(
-                scopeEntry = scopeEntry,
-                context = context,
-                locationString = candidate.info.usePath,
-                insertHandler = object : RsDefaultInsertHandler() {
-                    override fun handleInsert(
-                        element: RsElement,
-                        scopeName: String,
-                        context: InsertionContext,
-                        item: LookupElement
-                    ) {
-                        super.handleInsert(element, scopeName, context, item)
-                        if (RsCodeInsightSettings.getInstance().importOutOfScopeItems) {
-                            context.commitDocument()
-                            context.getElementOfType<RsElement>()?.let { candidate.import(it) }
+            val lookupElement = createLookupElementWithImportCandidate(scopeEntry, context, candidate)
+            result.addElement(lookupElement)
+        }
+    }
+
+    /** Adds completions for the case `HashMap::/*caret*/` where `HashMap` is not imported. */
+    private fun addCompletionsForOutOfScopeFirstPathSegment(path: RsPath, result: CompletionResultSet, context: RsCompletionContext) {
+        val qualifier = path.path
+        if (qualifier != null && qualifier.path == null && qualifier.typeQual == null && !qualifier.hasColonColon
+            && qualifier.resolveStatus == PathResolveStatus.UNRESOLVED) {
+            val importContext = ImportContext2.from(qualifier, ImportContext2.Type.AUTO_IMPORT) ?: return
+            val itemToCandidates = ImportCandidatesCollector2.getImportCandidates(importContext, qualifier.referenceName!!)
+                .groupBy { it.qualifiedNamedItem.item }
+            for ((_, candidates) in itemToCandidates) {
+                // Here all use path resolves to the same item, so we can just use the first of them
+                val firstUsePath = candidates.first().info.usePath
+
+                val newPath = RsCodeFragmentFactory(path.project).createPathInTmpMod(
+                    path.text,
+                    importContext.rootMod,
+                    importContext.pathInfo!!.pathParsingMode,
+                    path.allowedNamespaces(isCompletion = true),
```

Comment:
Please rebase on master and use `importContext.pathInfo.rootPathAllowedNamespaces`

Reply:
I suspect this may not be entirely correct. The completion can be invoked in the second section of the 3-segment path `foo::/*caret*::baz`, and in this case we must take into account `allowedNamespaces` and `pathParsingMode` of the 2nd segment